### PR TITLE
add global L2 projection for mixed 2D tensor splines

### PR DESCRIPTION
**Pull request by akva2**
Required for VTF output etc.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

File: src/ASM/ASMs2Dmx.C [hunk 1]
```diff
@@ -1026,3 +1027,112 @@ void ASMs2Dmx::getBoundaryNodes (int lIndex, IntVec& nodes, int basis,
...
+
+      int ip = (i2*ng1*nel1 + i1)*ng2;
+      for (int j = 0; j < ng2; j++, ip += ng1*(nel1-1))
+	for (int i = 0; i < ng1; i++, ip++)
+	{
+	  if (continuous)
+	    SplineUtils::extractBasis(spl1[ip],phi,dNdu);
+	  else
+	    phi = spl0[ip].basisValues;
+
```
Comment (kmokstad) [inline src/ASM/ASMs2Dmx.C]: `phi` is not used, so this statement can be skipped, and you can use phi2 as dummy in the extractBasis call such that you only need the `phi2` variable here.

File: src/ASM/ASMs2Dmx.C [hunk 2]
```diff
@@ -1026,3 +1027,112 @@ void ASMs2Dmx::getBoundaryNodes (int lIndex, IntVec& nodes, int basis,
...
+    surf->computeBasisGrid(gpar[0],gpar[1],spl0);
+
+  m_basis[0]->computeBasisGrid(gpar[0],gpar[1],spl01);
+
+  // Evaluate the secondary solution at all integration points
+  Matrix sField;
+  if (!this->evalSolution(sField,integrand,gpar.data()))
+  {
+    std::cerr <<" *** ASMs2D::assembleL2matrices: Failed for patch "<< idx+1
+	      <<" nPoints="<< gpar[0].size()*gpar[1].size() << std::endl;
```
Comment (kmokstad) [inline src/ASM/ASMs2Dmx.C]: fix indentation

File: src/ASM/ASMs2Dmx.C [hunk 3]
```diff
@@ -1026,3 +1027,112 @@ void ASMs2Dmx::getBoundaryNodes (int lIndex, IntVec& nodes, int basis,
...
+	  return false;
+	else if ((dA = 0.25*this->getParametricArea(1+iel)) < 0.0)
+	  return false; // topology error (probably logic error)
+      }
+
+      // --- Integration loop over all Gauss points in each direction ----------
+
+      int ip = (i2*ng1*nel1 + i1)*ng2;
+      for (int j = 0; j < ng2; j++, ip += ng1*(nel1-1))
+	for (int i = 0; i < ng1; i++, ip++)
```
Comment (kmokstad) [inline src/ASM/ASMs2Dmx.C]: fix indentation, one extra space?

File: src/ASM/ASMs2Dmx.C (hunk 2 above)
Comment (kmokstad) [inline src/ASM/ASMs2Dmx.C]: and also ASMs2D --> ASMs2Dmx in the error message

**Comment by akva2** [comment]
putting mixed code along the non-mixed feels more alien to me. but idc so sure. if that's what you want.

**Pull request merged**